esteid/idcard/README.md

```diff
@@ -30,21 +32,62 @@
    The container is saved to a desired location, and the temporary files are cleaned up.
    
 
-## Potential issues
+### Potential issues
 
-### Blocking on data processing
+#### Blocking on data processing
 
 Since the process no longer involves an external service, but the files to be signed can potentially
 be hefty, the process can block for a longer time while calculating hashes and (re-)packing the files
 into the container. (At the same time, employing an external service, it would block on a network read, 
 since the API is synchronous.)
 
-### Compatibility
+#### Compatibility
 
 Since the previous service-based action implementation provided for a very long inheritance tree
 of action classes which all override the `do_action` method in some way but still make a call to a
 non-existent `service`, it's advised that all those "actions" be scrapped in favor of a simpler API. 
 
-## The suggested API
+## Authentication with ID Card
+
+### Process Outline
+
+The request to the client to enter PIN is initialized by a specifically configured web server,
+with an option similar to `ssl_verify_client` on nginx.
+
+On nginx, it is only possible to protect an entire `server` section with this option, which means that
+in order to keep the site generally accessible to unauthenticated clients we need to set up a separate domain
+that would handle the ID card authentication process, and pass the client authentication data to the primary site
+via session or other means.
+
+There is a caveat: browsers keep a certificate in cache for an unspecified period of time,
+and a new request for PIN may not be triggered even once the authenticated ID card is removed or replaced.
+This is not possible to control by the web server. 
+
+Let's describe the key points of the ID card authentication process, based on the usual routine common for
+SmartID and MobileID. 
+
+The process thus should consist of the following steps:
+* While visiting the site (assuming `example.com` for this matter), the user clicks on a link to start the authentication process;
+* Instead of issuing a _start_ request to the backend (as in the usual routine), 
+  or a `hwcrypto` library call (as in the process of signing with ID card), the user is taken to a specifically configured
+  domain (e.g. `auth.example.com`) where they are asked for the ID Card's PIN code and 
+  proceed to allow the site to access the certificate;
+* the certificate is validated via OCSP;
+* the certificate or data obtained from it is saved to the session;
+* user is redirected to the protected page on the `example.com` site.
+
+The process above is by and large impractical. One improvement that could be made to it is to use an iframe instead of
```

Comment:
Not really sure I would have added iframes to the mix tbh. 

In etk I opened a new browser window which notified its parent about things with the same postMessage trick (window.open in original page and window.opener to use postMessage).

This way X-Frame-Options can be kept as `DENY` and CSP should not matter.

Reply:
What are your thoughts on the method without iframes?

Reply:
I feel it's not a very fine solution.
Window.open seems to be blocked by default (in Firefox, at least, it shows this yellow bar 'prevented a popup') and it's definitely not the best UX even if it weren't blocked, because it may be affected by window focusing preferences and there would be either tabs jumping or windows in the taskbar.
X-Frame-Options can still be kept as DENY for the main site, and the `auth` site would only need to enable one specific location, with everything else behind 404.
Also as per this TR https://www.w3.org/TR/CSP2/#frame-ancestors-and-frame-options , X-Frame-Options is called obsolete :)

Reply:
window.open is not blocked when called from a user event handler (like onclick)

Reply:
New tabs/windows are also used for Oauth flows so I don't really agree with your point about UX.

Reply:
> window.open is not blocked when called from a user event handler (like onclick)
forgot about it, thanks :)

For OAuth, when you need to enter data in a form, then it makes sense to open up a window, But a window that bears 0 significance to the user seems to me to be useless and not good. I also haven't seen any other services (LHV, Swedbank or Estonian government services) which would use a window.

And all in all, I don't really see any drawbacks to the iframe approach.

Reply:
the banks navigate you trough the id auth domain/location and then redirect you back. all in the same window. 

but this is nitpicking anyway and not too relevant